csrc/ops/alias.cpp

```diff
@@ -381,7 +381,17 @@ TensorView* transpose(TensorView* x) {
 
 // Padding widths are assumed to be non-negative. Currently there's no
 // validation.
-TensorView* pad(TensorView* inp, const std::vector<Val*>& pad_widths) {
+TensorView* pad(
+    TensorView* inp,
+    const std::vector<Val*>& pad_widths,
+    Val* value) {
```

Comment:
Do we want to have some check the data type of `value` match with the data type of `inp`? Not exactly sure how, though. It seems it should be fine if `value` can be cast to the data type of the `inp`.

Reply:
I think it may also make sense to insert `castOp` if the data type of `value` is different from the data type of `inp`. That way, the question about the type compatibility can be delegated to `castOp`.

Reply:
I think cast makes more sense.

FYI: This is the pytorch signature `- func: pad(Tensor self, SymInt[] pad, str mode="constant", float? value=None) -> Tensor `

Reply:
Yeah it will always insert a cast. The behavior is same as torch.

Reply:
> I think it may also make sense to insert `castOp` if the data type of `value` is different from the data type of `inp`. That way, the question about the type compatibility can be delegated to `castOp`.

I tried inserting a `castOp` on the value but it actually just discards the imaginary part of `value` and uses it as a `float`:
```c++
  float f13;
  f13 = (float)std::real(std::complex<double>(2.39999999999999991e+00,3.20000000000000018e+00));
  #pragma unroll 1
  for(nvfuser_index_t i20 = 0; i20 < (T0.size[0] + 2); ++i20) {
    if ((i20 < i158)) {
      T2[i20]
         = (((i20 - 1) >= 0) && ((i20 - 1) < T0.size[0])) ? T0[(T0.stride[0] * (i20 - 1))] : f13;
    }
  }
```
Since this seems unsafe and is not supported by PyTorch, I'm instead doing a `TORCH_CHECK` on the dtype of `value`: https://github.com/NVIDIA/Fuser/blob/57a9725b73d373eb3e0268597767fda3153a772e/csrc/ops/alias.cpp#L402-L404
If a user really wants the "discard imaginary" behavior they can always call `real()` on `value` manually when passing.

Reply:
I was thinking about inserting a cast `UnaryOp` when the value has a different type as the tensor. This way, you don't need to implement the type cast rules here just for `pad`.  Something like:

```
if (value->dtype() != dt) {
  value = castOp(dt, value);
}
```

should work. I assume we have the same type check for complex types already for `castOp`, so you don't need to do add that here, although that may provide a better error message.